0000-frontend-development-infrastructure-and-workflow-changes.md

```diff
@@ -0,0 +1,75 @@
+
+The tools in the MV* space are often classified as being opinionated or unopinionated with respect to the amount of flexibility they allow the developer. This often maps to the breadth of the issues the library or framework attempts to deal with.
+
+Reactjs is a component-based rendering engine. As such, it is both narrow in scope and unopinionated. This will facilitate integration with Rails with minimum side effects on the existing Ruby/Rails code base compared to the opinionated alternatives.
+
+React components demonstrate a high level of encapsulation and support dependency injection. As a result, the components are very testable. The implementation of the recommendations in this RFC will be accompanied by the addition of linting and testing to the client side development process.
+
+# Detailed design
+[design]: #detailed-design
+
+Reactjs uses javascript to compose components and render HTML. Data is passed to the component via props, which are immutable. Component state is mutable and managed by the component. When state changes the component is rerendered.
+
+Note: for those familiar with Angular 1, React components are similar to Angular directives.
+
+The [Fixes #13424 - WIP c3 patternfly react implementation PR] (https://github.com/theforeman/foreman/pull/3603) provides a POC for integrating Reactjs with foreman.
+
+Top level components added to an HTML page (mounted) with the `ReactDOM` package `render` method. This is the point at which the Rails view passes the baton to Reactjs.
+
+The paradigm for using React components in Rails views implemented in the POC is as follows:
+
+>The `mount_react_component` helper in the `ReactjsHelper` module is used in the Rails view to which the component is to be added. The helper has 3 paramaters: a string identifier for the specific component to be mounted, a jquery selector that is used to find the DOM element into which to inject the component and, optionally, initial data from the controller in JSON format. See *'app/views/statistics/index.html.erb'* in the POC for an example.
+
+``` erb
+    <div id="statistics"></div>
+         
+    <%= mount_react_component('StatisticsChartsList', '#statistics', @data)%>
+```
+
+> The helper injects inline javascript code into the Rails view which calls the `mount` method on the `reactMounter` object exposed on the global name space with the the parameters described above.
+
+> The `reactMounter` object manages a collection of outwardly facing components. The `mount` method finds the markup for the component specified by the identifier, adds the initial data (if provided) as props, finds the DOM element specified by the jquery selector and calls the `ReactDOM render` method with the component markup and a reference to the DOM element as parameters.
```

Comment:
Let me clarify what I mean by outwardly facing components. This is a top level component (basically a container) that is added to the HTML with ReactDOM.render. Components can be composed of other components and a component that is added to the HTML as part of another component in that components render method would not be considered outwardly facing.

The collection of outwardly facing is not necessarily limited to components written by us so data could be passed as props in the same way. That said, I expect that mounting 3rd party components directly into our HTML would be difficult to implement without encapsulation and single responsibility issues. My intuition suggests wrapping the 3rd party component in a component we write that would set up the connections to the app.

Let me also add that the above mentioned paradigm is a bit of an odd duck. foreman is not a single page app and it relies almost exclusively if not exclusively on server side rendering via erb. The purpose of the mounter is to allow us to go forward while not forcing us to make major changes in the structure and functioning of the app. In the more classic React case, most of the data would be retrieved via Ajax calls.